dipy/tracking/life.py

```diff
@@ -488,14 +522,191 @@ def fit(self, data, streamline, affine=None, evals=[0.001, 0, 0],
+        step_size : float
+            In optimization, the size of the gradient step change to the
+            parameters to perform in each round.
+        """
+        if sphere is None:
+            sphere = dpd.get_sphere()
 
+        signal_maker = self._signal_maker(evals=evals,
+                                          sphere=sphere)
+
+        if affine is None:
+            affine = np.eye(4)
+        else:
+            streamline = transform_streamlines(streamline, affine)
+
+        sl_as_coords = [np.round(s).astype(np.intp) for s in streamline]
+        cat_streamline = np.concatenate(sl_as_coords)
+        vox_coords = unique_rows(cat_streamline)
+        v2f = streamline_mapping(sl_as_coords, affine=np.eye(4))
 
-class FiberFit(ReconstFit):
+        (to_fit, weighted_signal, b0_signal, relative_signal, mean_sig,
+         vox_data) = self._fit_signals(data, vox_coords)
+        del weighted_signal, b0_signal, relative_signal, mean_sig
+        closest = {}
+        for sl_idx, ss in enumerate(streamline):
+            closest[sl_idx] = []
+            for node_idx in range(ss.shape[0]):
+                if node_idx == 0:
+                    g = ss[1] - ss[0]
+                elif node_idx == ss.shape[0]:
+                    g = ss[-1] - ss[-2]
+                else:
+                    g = ss[node_idx] - ss[node_idx-1]
+                closest[sl_idx].append(sphere.find_closest(g))
+        # We only consider the diffusion-weighted signals in fitting:
+        n_bvecs = self.gtab.bvals[~self.gtab.b0s_mask].shape[0]
+        beta = np.zeros(len(streamline))
+        range_bvecs = np.arange(n_bvecs).astype(np.intp)
+
+        # Optimization-related stuff:
+        iteration = 0
+        ss_residuals_min = np.inf
+        sse_best = np.inf
+        error_checks = 0  # How many error checks have we done so far
+        y_hat = np.zeros(to_fit.shape)
+
+        # Cache some facts about relationship between voxels and fibers:
+        max_s_per_vox = np.max([len(t) for t in v2f.values()])
+        s_in_vox = {}
+        for v_idx in range(vox_coords.shape[0]):
+            s_in_vox[v_idx] = []
+            mat_row_idx = (range_bvecs + v_idx * n_bvecs).astype(np.intp)
+            this_vox = (vox_coords[v_idx][0], vox_coords[v_idx][1],
+                        vox_coords[v_idx][2])
+            # For each fiber in that voxel:
+            for sl_idx in v2f[this_vox]:
+                find_vox = np.logical_and(
+                  np.logical_and(
+                    sl_as_coords[sl_idx][:, 0] == vox_coords[v_idx][0],
+                    sl_as_coords[sl_idx][:, 1] == vox_coords[v_idx][1]),
+                  sl_as_coords[sl_idx][:, 2] == vox_coords[v_idx][2])
+                nodes_in_vox = np.where(find_vox)[0]
+                s_in_vox[v_idx].append((sl_idx, nodes_in_vox))
+
+        # We no longer need these variables:
+        del v2f, streamline, sl_as_coords
+
+        delta = np.zeros(beta.shape)
+        while 1:
```

Comment:
As this part here seem to do fixed stepsize gradient descent, you might want to try http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.optimize.nnls.html to see if its faster in case. If not, the problem is convex, so you might wan tot try an already made solver also for ease of maintenance.